cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -301,26 +301,36 @@ Result<HdfsOptions> HdfsOptions::FromUri(const Uri& uri) {
     options_map.emplace(kv.first, kv.second);
   }
 
-  const auto port = uri.port();
-  if (port == -1) {
-    options.ConfigureEndPoint(uri.host(), kDefaultHdfsPort);
-  } else {
-    options.ConfigureEndPoint(uri.host(), port);
-  }
-
+  auto useHdfs3 = false;
   auto it = options_map.find("use_hdfs3");
   if (it != options_map.end()) {
     const auto& v = it->second;
     if (v == "1") {
-      options.ConfigureHdfsDriver(true);
+      options.ConfigureHdfs3Driver(true);
+      useHdfs3 = true;
     } else if (v == "0") {
-      options.ConfigureHdfsDriver(false);
+      options.ConfigureHdfs3Driver(false);
     } else {
       return Status::Invalid(
           "Invalid value for option 'use_hdfs3' (allowed values are '0' and '1'): '", v,
           "'");
     }
   }
+
+  std::string host;
+  if (useHdfs3) {
+    host = uri.host();
```

Comment:
Does this mean libhdfs3 doesn't support `viewfs`? If so, perhaps return an error in that case?

Reply:
Yes, it doesn't. We actually even have a Python implementation on top of https://github.com/dask/hdfs3 to do the viewfs resolution with hdfs3 on our own. But as this project is deprecated and there is a new hdfs3 driver for Hadoop 3.2+ not clear where to upstream this.

Why returning an error ? The resolution will work fine for hdfs with hdfs3 driver. It will just work as before for hdfs3 driver.

Reply:
Hmm, what do you mean with "the resolution will work fine"? If I ask for `viewfs://something`, it silently downgrades to `hdfs://something`? Is that desirable?

Reply:
Yes, for driver libhdfs3: viewfs://something => will become hdfs://something
This works if the namenode is existing in the config which is the case on our config. We only have issues for cross cluster access and if we use hdfs delgation tokens where the delegation tokens are asked with viewfs resolution -> we get the delegation token for hdfs://the-real-name-node but then pyarrow tries to resolve hdfs://something and the delegation token is missing
Also 'default' will keep working, default will get resolved to the default namenode which can be also viewfs and will therefore properly resolved on hdfs side.

As I said it will be as before for hdfs3. No regression and implementing this for hdfs3 driver would be too much work and obsolete with the new hdfs3 driver.
On ourside for the moment we plan using libhdfs only and see if we switch to new libhdfs3 implementation later.

Reply:
So it depends on the user's config. I think it would be safer to fail with an explicit message (such as "viewfs is not supported with the hdfs3 driver").

Reply:
What about a warning message ? 
> viewfs://namenode resolves hdfs://namenode with hdfs3 driver
Not sure it is clearer but I think it currently works in 80% of the cases (at least in 80% of the cases on our cluster) and people might write code that also uses pyarrow with libhdfs3 on our cluster.
Not sure we should break people's code as they might already use viewfs even with the existing implementation. 

TensorFlow also fails for viewfs:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/hadoop/hadoop_file_system.cc#L151
Meaning viewfs only works if viewfs is the default driver which is somehow what you suggest.

Reply:
I'm fine with a warning. You can use `ARROW_LOG(WARNING) << ...` for that.

Reply:
OK. Done